File: src/main/java/org/tobi29/scapes/engine/gui/GuiComponentHeavy.java
package org.tobi29.scapes.engine.gui;

import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.opengl.GL;
import org.tobi29.scapes.engine.opengl.VAO;
import org.tobi29.scapes.engine.opengl.matrix.Matrix;
import org.tobi29.scapes.engine.opengl.matrix.MatrixStack;
import org.tobi29.scapes.engine.opengl.shader.Shader;
import org.tobi29.scapes.engine.opengl.texture.Texture;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.ThreadLocalUtil;
import org.tobi29.scapes.engine.utils.Triple;
import org.tobi29.scapes.engine.utils.math.vector.Vector2;
import org.tobi29.scapes.engine.utils.math.vector.Vector2d;
import org.tobi29.scapes.engine.utils.math.vector.Vector3;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

public abstract class GuiComponentHeavy extends GuiComponent {
    private static final ThreadLocal<GuiRenderer> RENDERER =
            ThreadLocalUtil.of(GuiRenderer::new);
    private final AtomicBoolean dirty = new AtomicBoolean(true);
    protected List<Pair<VAO, Texture>> meshes = Collections.emptyList();
    private Vector2 lastSize = Vector2d.ZERO;
    private boolean hasHeavyChild;

    protected GuiComponentHeavy(GuiLayoutData parent) {
        super(parent);
    }

    @Override
    protected void render(GL gl, Shader shader, Vector2 size, double delta) {
        if (visible) {
            MatrixStack matrixStack = gl.matrixStack();
            Matrix matrix = matrixStack.push();
            transform(matrix, size);
            if (dirty.getAndSet(false) || !lastSize.equals(size)) {
                GuiRenderer renderer = RENDERER.get();
                hasHeavyChild = render(renderer, size);
                meshes = renderer.finish();
                lastSize = size;
            }
            Streams.forEach(meshes, mesh -> {
                mesh.b.bind(gl);
                mesh.a.render(gl, shader);
            });
            renderComponent(gl, shader, size, delta);
            if (hasHeavyChild) {
                GuiLayoutManager layout = layoutManager(size);
                for (Triple<GuiComponent, Vector2, Vector2> component : layout
                        .layout()) {
                    Vector3 pos = applyTransform(-component.b.doubleX(),
                            -component.b.doubleY(), size);
                    if (-pos.doubleX() >= -component.c.doubleX() &&
                            -pos.doubleY() >= -component.c.doubleY() &&
                            -pos.doubleX() <= size.doubleX() &&
                            -pos.doubleY() <= size.doubleY()) {
                        Matrix childMatrix = matrixStack.push();
                        childMatrix.translate(component.b.floatX(),
                                component.b.floatY(), 0.0f);
                        component.a.render(gl, shader, component.c, delta);
                        matrixStack.pop();
                    }
                }
            }
            matrixStack.pop();
        }
    }

    @Override
    protected void renderOverlays(GL gl, Shader shader) {
        super.renderOverlays(gl, shader);
        if (visible) {
            renderOverlay(gl, shader);
        }
    }

    @Override
    protected boolean renderLightweight(GuiRenderer renderer, Vector2 size) {
        return true;
    }

    @Override
    protected boolean render(GuiRenderer renderer, Vector2 size) {
        boolean hasHeavy = false;
        MatrixStack matrixStack = renderer.matrixStack();
        updateMesh(renderer, size);
        GuiLayoutManager layout = layoutManager(size);
        for (Triple<GuiComponent, Vector2, Vector2> component : layout
                .layout()) {
            Matrix childMatrix = matrixStack.push();
            childMatrix.translate(component.b.floatX(), component.b.floatY(),
                    0.0f);
            hasHeavy |= component.a.renderLightweight(renderer, component.c);
            matrixStack.pop();
        }
        return hasHeavy;
    }

    @Override
    protected void update(ScapesEngine engine, double delta) {
        super.update(engine, delta);
        if (visible) {
            updateComponent(engine, delta);
        }
        parent.parent().ifPresent(GuiComponent::activeUpdate);
    }

    @Override
    public void dirty() {
        dirty.set(true);
    }

    protected void updateComponent(ScapesEngine engine, double delta) {
    }

    protected void renderComponent(GL gl, Shader shader, Vector2 size,
            double delta) {
    }

    protected void renderOverlay(GL gl, Shader shader) {
    }
}

File: Engine/src/main/kotlin/org/tobi29/scapes/engine/gui/GuiComponentHeavy.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.gui

import org.tobi29.scapes.engine.ScapesEngine
import org.tobi29.scapes.engine.graphics.*
import org.tobi29.scapes.engine.utils.AtomicBoolean
import org.tobi29.scapes.engine.utils.ThreadLocal
import org.tobi29.scapes.engine.utils.math.vector.Vector2d

abstract class GuiComponentHeavy : GuiComponent {
    private val dirty = AtomicBoolean(true)
    protected var meshes: List<Pair<Model, Texture>>? = null
    private var lastSize = Vector2d.ZERO
    private var lastPixelSize = Vector2d.ZERO
    private var hasHeavyChild = false

    constructor(
            parent: GuiLayoutData
    ) : super(parent)

    internal constructor(
            engine: ScapesEngine,
            parent: GuiLayoutData
    ) : super(engine, parent)

    override fun render(gl: GL,
                        shader: Shader,
                        size: Vector2d,
                        pixelSize: Vector2d,
                        delta: Double) {
        if (visible) {
            gl.matrixStack.push { matrix ->
                transform(matrix, size)
                if (dirty.getAndSet(false) || lastSize != size ||
                        lastPixelSize != pixelSize) {
                    lastPixelSize = pixelSize
                    val renderer = RENDERER.get()
                    renderer.pixelSize = pixelSize
                    hasHeavyChild = render(renderer, size)
                    meshes = renderer.finish()
                    lastSize = size
                }
                meshes?.forEach {
                    it.second.bind(gl)
                    it.first.render(gl, shader)
                }
                renderComponent(gl, shader, size, pixelSize, delta)
                if (hasHeavyChild) {
                    val layout = layoutManager(size)
                    for ((component, position, childSize) in layout.layout()) {
                        val pos = applyTransform(-position.x, -position.y, size)
                        if (-pos.x >= -childSize.x && -pos.y >= -childSize.y
                                && -pos.x <= size.x && -pos.y <= size.y) {
                            gl.matrixStack.push { childMatrix ->
                                childMatrix.translate(position.floatX(),
                                        position.floatY(), 0.0f)
                                component.render(gl, shader, childSize,
                                        pixelSize, delta)
                            }
                        }
                    }
                }
            }
        }
    }

    public override fun renderOverlays(gl: GL,
                                       shader: Shader,
                                       pixelSize: Vector2d) {
        super.renderOverlays(gl, shader, pixelSize)
        if (visible) {
            renderOverlay(gl, shader, pixelSize)
        }
    }

    override fun renderLightweight(renderer: GuiRenderer,
                                   size: Vector2d): Boolean {
        return true
    }

    override fun render(renderer: GuiRenderer,
                        size: Vector2d): Boolean {
        var hasHeavy = false
        val matrixStack = renderer.matrixStack()
        updateMesh(renderer, size)
        val layout = layoutManager(size)
        for ((component, position, childSize) in layout.layout()) {
            matrixStack.push { childMatrix ->
                childMatrix.translate(position.floatX(), position.floatY(),
                        0.0f)
                hasHeavy = hasHeavy or component.renderLightweight(renderer,
                        childSize)
            }
        }
        return hasHeavy
    }

    override fun update(delta: Double) {
        super.update(delta)
        if (visible) {
            updateComponent(delta)
        }
        parent.parent?.activeUpdate()
    }

    override fun dirty() {
        dirty.set(true)
    }

    protected open fun updateComponent(delta: Double) {
    }

    protected open fun renderComponent(gl: GL,
                                       shader: Shader,
                                       size: Vector2d,
                                       pixelSize: Vector2d,
                                       delta: Double) {
    }

    protected open fun renderOverlay(gl: GL,
                                     shader: Shader,
                                     pixelSize: Vector2d) {
    }

    companion object {
        private val RENDERER = ThreadLocal { GuiRenderer() }
    }
}
